servicetalk-http-api/src/main/java/io/servicetalk/http/api/HeaderUtils.java

```diff
@@ -667,12 +667,17 @@ static boolean hasContentType(final HttpHeaders headers,
             return false;
         }
 
-        if (UTF_8.equals(expectedCharset) &&
-                ((contentEqualsIgnoreCase(expectedContentType, TEXT_PLAIN) &&
-                        contentEqualsIgnoreCase(contentTypeHeader, TEXT_PLAIN_UTF_8)) ||
-                (contentEqualsIgnoreCase(expectedContentType, APPLICATION_X_WWW_FORM_URLENCODED) &&
-                        contentEqualsIgnoreCase(contentTypeHeader, APPLICATION_X_WWW_FORM_URLENCODED_UTF_8)))) {
-            return true;
+        if (UTF_8.equals(expectedCharset)) {
```

Comment:
It looks like what is required is:

```java
if (UTF_8.equals(expectedCharset)) {
    if (!hasCharset(contentTypeHeader)) {
        return true;
    }
    if (contentEqualsIgnoreCase(expectedContentType, TEXT_PLAIN) &&
            contentEqualsIgnoreCase(contentTypeHeader, TEXT_PLAIN_UTF_8)) {
        return true;
    }
    if (contentEqualsIgnoreCase(expectedContentType, APPLICATION_X_WWW_FORM_URLENCODED) &&
            contentEqualsIgnoreCase(contentTypeHeader, APPLICATION_X_WWW_FORM_URLENCODED_UTF_8)) {
        return true;
    }
}
```

I would also suggest adding another test to make sure we are checking for content-type equality without charset:

```java
assertFalse(HeaderUtils.hasContentType(
        headersWithContentType(of("image/png")), APPLICATION_X_WWW_FORM_URLENCODED, UTF_8));
```

Reply:
@NiteshKant applied to any content type and added tests.